src/dfx/src/lib/builders/mod.rs

```diff
@@ -324,27 +324,49 @@ fn ensure_trailing_newline(s: String) -> String {
     }
 }
 
-pub fn run_command(args: Vec<String>, vars: &[Env<'_>], cwd: &Path) -> DfxResult<()> {
-    let (command_name, arguments) = args.split_first().unwrap();
-    let canonicalized = dfx_core::fs::canonicalize(&cwd.join(command_name))
-        .or_else(|_| which::which(command_name))
-        .map_err(|_| anyhow!("Cannot find command or file {command_name}"))?;
-    let mut cmd = Command::new(canonicalized);
-
-    cmd.args(arguments)
-        .current_dir(cwd)
-        .stdout(Stdio::inherit())
-        .stderr(Stdio::inherit());
-
+/// Run a command and return its output.
+/// If the io_inherit is true, the return bytes will always be empty.
+pub fn run_command(
+    command: &str,
+    vars: &[Env<'_>],
+    cwd: &Path,
+    io_inherit: bool,
+) -> DfxResult<Vec<u8>> {
+    // No commands, noop.
+    if command.is_empty() {
+        return Ok(vec![]);
+    }
+    let (first, rest) = match command.split_once(' ') {
```

Comment:
Since this no longer calls shell_words::split, it won't handle spaces in the command name. (Of course, no one would do this, but the call to shell_words::split handled this before, so this could be a breaking change)
```
"build": "command\\ with\\ spaces the parameters" 
```
or
```
"build": "./some\\ commands/something the parameters."
```

```

Reply:
I refactored this part a bit.

After simplification, there are two cases for the `command`:
- It's a file name relative to `cwd`, which may contains whitespace
  - canonicalize the path and execute it directly.
- It's a shell command which may have arguments
  - simply run `sh -c <command>` which allow pipes